src/backend/distributed/worker/worker_shard_visibility.c

```diff
@@ -257,72 +266,168 @@ RelationIsAKnownShard(Oid shardRelationId)
+}
+
+
+/*
+ * ShouldHideShardsFromApplication determines whether we should hide shards based on
+ * the current application name.
  */
 static bool
-ReplaceTableVisibleFunctionWalker(Node *inputNode)
+ShouldHideShardsFromApplication(void)
 {
-	if (inputNode == NULL)
+	List *prefixList = NIL;
+
+	/* SplitGUCList scribbles on the input */
+	char *splitCopy = pstrdup(HideShardsFromAppNamePrefixes);
+
+	if (!SplitGUCList(splitCopy, ',', &prefixList))
 	{
+		/* invalid GUC value, ignore */
 		return false;
 	}
 
-	if (IsA(inputNode, FuncExpr))
+	char *appNamePrefix = NULL;
+	foreach_ptr(appNamePrefix, prefixList)
 	{
-		FuncExpr *functionToProcess = (FuncExpr *) inputNode;
-		Oid functionId = functionToProcess->funcid;
-
-		if (functionId == PgTableVisibleFuncId())
+		/* always hide shards when one of the prefixes is * */
+		if (strcmp(appNamePrefix, "*") == 0)
 		{
-			/*
-			 * We simply update the function id of the FuncExpr for
-			 * two reasons: (i) We don't want to interfere with the
-			 * memory contexts so don't want to deal with allocating
-			 * a new functionExpr (ii) We already know that both
-			 * functions have the exact same signature.
-			 */
-			functionToProcess->funcid = CitusTableVisibleFuncId();
+			return true;
+		}
 
-			/* although not very likely, we could have nested calls to pg_table_is_visible */
-			return expression_tree_walker(inputNode, ReplaceTableVisibleFunctionWalker,
-										  NULL);
+		/* compare only the first first <prefixLength> characters */
+		int prefixLength = strlen(appNamePrefix);
+		if (strncmp(application_name, appNamePrefix, prefixLength) == 0)
+		{
+			return true;
 		}
 	}
-	else if (IsA(inputNode, Query))
+
+	return false;
+}
+
+
+/*
+ * FilterShardsFromPgclass adds a NOT relation_is_a_known_shard(oid) filter
+ * to the security quals of pg_class RTEs.
+ */
+static bool
+FilterShardsFromPgclass(Node *node, void *context)
+{
+	if (node == NULL)
 	{
-		return query_tree_walker((Query *) inputNode, ReplaceTableVisibleFunctionWalker,
-								 NULL, 0);
+		return false;
 	}
 
-	return expression_tree_walker(inputNode, ReplaceTableVisibleFunctionWalker, NULL);
+	if (IsA(node, Query))
+	{
+		Query *query = (Query *) node;
+		MemoryContext queryContext = GetMemoryChunkContext(query);
+
+		/*
+		 * We process the whole rtable rather than visiting individual RangeTblEntry's
+		 * in the walker, since we need to know the varno to generate the right
+		 * fiter.
+		 */
+		int varno = 0;
+		RangeTblEntry *rangeTableEntry = NULL;
+
+		foreach_ptr(rangeTableEntry, query->rtable)
+		{
+			varno++;
+
+			if (rangeTableEntry->rtekind != RTE_RELATION ||
+				rangeTableEntry->relid != RelationRelationId)
+			{
+				/* not pg_class */
+				continue;
+			}
+
+			/* make sure the expression is in the right memory context */
+			MemoryContext originalContext = MemoryContextSwitchTo(queryContext);
+
+			/* add NOT relation_is_a_known_shard(oid) to the security quals of the RTE */
+			rangeTableEntry->securityQuals =
```

Comment:
> what if there is already a security qual (e.g., is list_make1 safe enough )?

You cannot create policies on catalog tables.

> trade-off compared to adding the filter like we do in other places such as?

Quals in the join tree have different/incorrect semantics, since they are evaluated after outer joins.

Reply:
> I believe this should be fine for INSERT..SELECT pushdown filters we add?

Whether it's "fine" is a bit of a philosophical question. It changes the behaviour since we're changing the semantics of the INSERT..SELECT to exclude non-matching rows.

e.g.
```
create table test (x int, y int);
create table test2 (x int, y int);
insert into test2 values (1,2);
insert into test select b.* from test2 a left join test b on (a.x = b.x);
INSERT 0 1
select create_distributed_table('test','x');
select create_distributed_table('test2','x');
insert into test select b.* from test2 a left join test b on (a.x = b.x);
INSERT 0 0
```
on the other hand we do not allow NULL distribution column in distributed tables, so the alternative would be to fail